### PR TITLE
Adding FetchTime and QueryTime for each Query

### PR DESCRIPTION
**Pull request by Tushar3099**
Fixes Issue: #295 

**Type Of Change**
----

- [x] New feature (non-breaking change which adds functionality)

**Changes**
----

- Modified  `QueryResponse` to send queryTime in Nanoseconds.
- Refactored some functions in `Matrix.tsx`
- Calculated fetchTime for fetching query.
- Modifying HTML structure and styling for showing URL.
- Added fetchTime and queryTime in the top-right corner.

**Preview**
----

https://user-images.githubusercontent.com/51849085/111930492-16376000-8adf-11eb-84af-530bc0aa3052.mp4

### CODE REVIEW

File: dashboard/v1.1/src/pages/Monitoring/RouteDetails.tsx
Comment (aquibbaig) [inline dashboard/v1.1/src/pages/Monitoring/RouteDetails.tsx:153]: Is this PR ready for review? Please request reviews once it is and remove comments before doing that

Comment (Tushar3099) [reply dashboard/v1.1/src/pages/Monitoring/RouteDetails.tsx:153]: Thanks for the reply @aquibbaig, will do the prescribed changes and request for review.

**Comment by Tushar3099** [comment]
@Harkishen-Singh @aquibbaig, my PR is ready for review. Please review it.

File: tsdb/querier/querier.go
Comment (Harkishen-Singh) [inline tsdb/querier/querier.go:137]: I wouldn't do this here. The way this should be done is, at line 103, we must use a defer so that the only statement works for all cases and we make sure that there are no edge cases.

```golang
defer func() {
	q.queryResponse.TimeInvolved = time.Since(stamp).Nanoseconds()
}()
```

Comment (Harkishen-Singh) [reply tsdb/querier/querier.go:137]: And remove all other places for this calculation, as this itself is enough.

Note: Query time is the time from when you enter this func to exit of this func, since that's where the actual querying takes place. All other times fall under fetch time.

File: dashboard/v1.1/src/pages/Monitoring/RouteDetails.tsx
Comment (Harkishen-Singh) [inline dashboard/v1.1/src/pages/Monitoring/RouteDetails.tsx:162]: 2 points:
1. What happens if the time is never in ms or micros? If it's in seconds, then I guess this will show some thousand ms which is never good.
2. I will be happy if this is move concise. A bold query-time is never good. Let's make it compact and cover the timing with a beautify badge that can be green in colour.

File: dashboard/v1.1/src/pages/Monitoring/RouteDetails.tsx
Comment (Harkishen-Singh) [inline dashboard/v1.1/src/pages/Monitoring/RouteDetails.tsx:58]: Can we not have this as an object? using an array for this purpose is weird.

File: dashboard/v1.1/src/utils/queryTypes.ts
Comment (Harkishen-Singh) [inline dashboard/v1.1/src/utils/queryTypes.ts:119]:
```suggestion
  evaluationTime: number;
```

File: tsdb/querier/querier.go
Comment (Harkishen-Singh) [reply tsdb/querier/querier.go:137]: In order to save from computations in UI, How about using `time.Since(stamp).String()`?

**Review by Harkishen-Singh — CHANGES_REQUESTED** [review]
Some questions:
1. I need to double-check, but I think we just query once a metric. That means, the query & fetch timings should never change even when the section is changed.

Please let me know if I am wrong.

**Comment by Harkishen-Singh** [comment]
@aquibbaig can you please give a quick review of the react code?

**Comment by Tushar3099** [comment]
> Some questions:
> 
> 1. I need to double-check, but I think we just query once a metric. That means, the query & fetch timings should never change even when the section is changed.
> 
> I am happy to be wrong. Please point me out.

@Harkishen-Singh Actually inside SendMatrix function, I observed that : 
```
		go parallelQueryExec(matrix.PingChain.Path, querier.TypeFirst, curr, from, chans[0])
		go parallelQueryExec(matrix.JitterChain.Path, querier.TypeFirst, curr, from, chans[1])
		go parallelQueryExec(matrix.MonitorChain.Path, querier.TypeFirst, curr, from, chans[2])
		matrixResponse = map[string]querier.QueryResponse{
			"ping":    <-chans[0],
			"jitter":  <-chans[1],
			"monitor": <-chans[2],
		}
```
each of the queries was done parallelly in the different channels that took different time durations to complete so I assumed it will be different for each section. Please correct me if I misunderstood something.
Also, I want to ask about the `fetchTime` implementation that I did in the frontend. I actually calculated the time duration for the API to return data and assigned it to `fetchTime`. Is that implementation correct?

File: tsdb/querier/querier.go
Comment (Tushar3099) [reply tsdb/querier/querier.go:137]: - Absolutely right, `defer` will be best for the time duration calculation here. I will implement it that way.
- Yes, the `time.Since(stamp).String()` will help us avoid the calculations in the frontend, also then the return type will be a string I assume. I had a question, sending time as a string will prevent us from doing any calculation on the time in the frontend, that shouldn't be a problem, right?

File: dashboard/v1.1/src/pages/Monitoring/RouteDetails.tsx
Comment (Tushar3099) [reply dashboard/v1.1/src/pages/Monitoring/RouteDetails.tsx:162]: - I can use the moment library to handle the time symbols. That should solve the issue. May I use that?
- I will surely do the prescribed styling changes.

File: dashboard/v1.1/src/pages/Monitoring/RouteDetails.tsx
Comment (Tushar3099) [reply dashboard/v1.1/src/pages/Monitoring/RouteDetails.tsx:58]: My bad, an object struct will be a better fit for the use case. I will do the changes.

File: tsdb/querier/querier.go
Comment (Harkishen-Singh) [reply tsdb/querier/querier.go:137]: I don't think there are any calculations involved with time in the frontend, so this should be fine.

File: dashboard/v1.1/src/pages/Monitoring/RouteDetails.tsx
Comment (Harkishen-Singh) [reply dashboard/v1.1/src/pages/Monitoring/RouteDetails.tsx:162]: point 1 will be fixed if we follow https://github.com/bench-routes/bench-routes/pull/376#discussion_r598098250.

**Comment by Harkishen-Singh** [comment]
> queries was done parallelly in the different channels that took different time durations to complete so I assumed it will be different for each section.

Ah, I missed that point. That makes the implementation correct in monitoring screen.

BTW, I don't see the code for query and fetch timings in all other graphs. Are they still under WIP? @Tushar3099 

PS: there are compilation errors that needs a fix.

File: dashboard/v1.1/src/pages/Monitoring/RouteDetails.tsx
Comment (Tushar3099) [reply dashboard/v1.1/src/pages/Monitoring/RouteDetails.tsx:162]: Yeah, using `time.String()` will fix the evaluation time automatically. I was asking about the fetch time, which I am calculating in the frontend because that will give me result in milliseconds : 
```
      const start = performance.now();
      const details = await fetchDetails(instance);
      const end = performance.now();
```

**Comment by Tushar3099** [comment]
Yes, @Harkishen-Singh the last commit was done by mistake, right now I am implementing all the prescribed changes.
Do I have to show the query and fetch time in any other place also rather than the Routes Detail Screen? Please tell me if I overlooked something.

File: dashboard/v1.1/src/pages/Monitoring/RouteDetails.tsx
Comment (Harkishen-Singh) [reply dashboard/v1.1/src/pages/Monitoring/RouteDetails.tsx:162]: Yeah, you should make sure that all units are handled

**Comment by Harkishen-Singh** [comment]
Query and fetch times would be shown in all screens that have a graph which are being fetched through the API.

For example, dashboard has 7 graphs in total, but we should have only 2, the first for syste metrics and second for jounrla metrics. The test screen will also have a similar story (single fetch & query time on each dropdown click).

@Tushar3099

File: dashboard/v1.1/src/pages/Monitoring/Matrix.tsx
Comment (aquibbaig) [inline dashboard/v1.1/src/pages/Monitoring/Matrix.tsx:77]: `fetchtime` is not a prop of type `RouteDetails`.

File: dashboard/v1.1/src/pages/Monitoring/Matrix.tsx
Comment (aquibbaig) [inline dashboard/v1.1/src/pages/Monitoring/Matrix.tsx:54]: NIT: Let's add the return type (in other places too)
```diff
- const fetchDetails = async (instance: TimeSeriesPath)
+ const fetchDetails = async (instance: TimeSeriesPath): Promise<RouteDetails>
```

File: dashboard/v1.1/src/pages/Monitoring/RouteDetails.tsx
Comment (aquibbaig) [reply dashboard/v1.1/src/pages/Monitoring/RouteDetails.tsx:58]: Please add a comment on why `queryTime` has a structure as you have mentioned (e.g. ping, jitter, etc.)

**Review by aquibbaig — CHANGES_REQUESTED** [review]
Great work! This PR is looking good, with small changes here and there. From a design perspective, will it be better to show something like "query returned in x ms"? That would depict our actions better as providing positive user feedback.

**Comment by Tushar3099** [comment]
@Harkishen-Singh @aquibbaig  I am making all the prescribed changes currently and will make the commit ASAP.

**Comment by Tushar3099** [comment]
@Harkishen-Singh @aquibbaig  I tried to implement all the prescribed changes up to my best. Right now, `fetchTime` and `queryTime` are only shown in the Monitored Routes screen. Showing them on the Dashboard screen and Test screen is still WIP.

I have updated the Preview to show the current changes.

**Comment by Harkishen-Singh** [comment]
Short nit: when we initially click on the route, and the response length tab is first which is opened, why don't the query time gets shown there? I see only the fetch time being shown.

**Comment by Tushar3099** [comment]
@Harkishen-Singh Actually we are getting the queryTime as : 
```
              {
                    ping : "string",
                    jitter : "string",
                    monitor : "string"
              }
```
So I am showing the respective times in respective sections : 
```
            queryTime(ping) --> Ping Section
            queryTime(jitter) --> Jitter Section
            queryTime(monitor) --> Response Delay Section
```
I was not sure which `queryTime` to show in the ResponseLength Section so I omitted queryTime there. Please pardon my mistake and guide what will be the right thing to do.

File: tsdb/querier/querier.go
Comment (Harkishen-Singh) [inline tsdb/querier/querier.go:114]:
```suggestion
			if i, ok = json.Marshal(res); !ok {
			    log.Error("invalid type received")
			}
```

we should log error if not ok is false

**Review by Harkishen-Singh — APPROVED** [review]
A small change would have been better, but considering this will be updated in binary tsdb, I wouldn't block this. Thank you very much.

**Comment by Harkishen-Singh** [comment]
@Tushar3099 can you please address my last comment and resolve the commit?

File: tsdb/querier/querier.go
Comment (Harkishen-Singh) [inline tsdb/querier/querier.go:120]: This can be written as.

```suggestion
	defer func() {
		var res QueryResponse
		if !q.encoding {
		        res, ok := i.(QueryResponse)
		        if !ok {
		           log.Error("type 'QueryResponse' not received")
		        }
			res.EvaluationTime = time.Since(stamp).String()
			i = res
			return
		}
		err := json.Unmarshal(i.([]byte), &res)
		if err != nil {
			log.Errorln(fmt.Errorf("Decoding error: %s", err.Error()).Error())
		}
		res.EvaluationTime = time.Since(stamp).String()
		i, ok = json.Marshal(res)
		if !ok {
		      log.Error("invalid marshalling: ", err.Error())
	        } 
	}()
```

**Comment by Harkishen-Singh** [comment]
Let me know when this is ready for review

**Comment by Tushar3099** [comment]
Yeah, I will update soon, actually, there are some bugs with the recent changes. I am resolving them and will update the PR asap.

**Comment by Tushar3099** [comment]
@Harkishen-Singh, there are many conflicts between this PR and #364, as both have the same working files, many changes are overwriting each other, so it will take some time to resolve and update. Will try to do it asap.

**Comment by Tushar3099** [comment]
@Harkishen-Singh @aquibbaig I have merged my PR with #364. Please review it.

**Review by Harkishen-Singh — APPROVED** [review]
LGTM! Thank you for this. However, this will have to wait before #439 because of some performance issues. Sorry for the inconvenience. I will merge once that PR is merged.

**Comment by Tushar3099** [comment]
@Harkishen-Singh, I can revert to the last commit so that it won't have any dependency on #439, and then we can merge it normally because my PR has no dependency in itself with #439.

**Comment by Harkishen-Singh** [comment]
Let's wait for a day or two. We are testing the that commit. If it seems fine, we won't revert then. So better to hold on at present.

**Comment by Tushar3099** [comment]
@Harkishen-Singh @aquibbaig Please review and merge my PR as the revert of PR #439 is closed, for which my PR was put on hold.

**Comment by Tushar3099** [comment]
@Harkishen-Singh Can you please review and merge this PR? Thank you

**Review by Harkishen-Singh — APPROVED** [review]
Thank you for this. Just a note to myself, in the new UI, the timings can be made more compact.

**Pull request merged**
